src/common/linear_assignment_optimizer.h

```diff
@@ -156,3 +158,129 @@ void auctionOptimization(M &value_matrix, std::array<int, N> &assignment, int si
+					if (label != *part_it_) { return false; }
+					part_it_++;
+				}
+			}
+			for (int i = 0; i < source_slice.size(); i++) {
+				detail::Slice::PartProxy part = source_slice[i];
+				detail::Slice::PartProxy part_ = source_slice_[i];
+				if (part.size() != part_.size()) { return false; }
+				auto part_it_ = part_.begin();
+				for (const auto &label : part) {
+					if (label != *part_it_) { return false; }
+					part_it_++;
+				}
+			}
+
+			// match is good
+			memcpy(assignemt.data(), assignemt_.data(),
+			       sizeof(int) * target_slice.size());
+
+			return true;
+		}
+
+		template <std::size_t N>
+		void store(Goal::Slice &target_slice, Goal::Slice &source_slice,
+		           std::array<int, N> &assignemt) {
+			target_slice_ = target_slice;
+			source_slice_ = source_slice;
+			assignemt_.resize(target_slice_.size());
+			for (int i = 0; i < target_slice_.size(); i++) {
+				assignemt_[i] = assignemt[i];
+			}
+		}
+
+	protected:
+		Goal::Slice target_slice_, source_slice_;
+		std::vector<int> assignemt_;
+	};
+
+	using EntriesContainer = std::vector<OptimizerInputOutput>;
+	using HashType = int64_t;
+
+	LinearAssignmentCache() : cachedData_(kModSlice_) {}
+
+	HashType getLabelValue(MediaLabel label) {
+		return static_cast<MediaLabelManager::HandleValue>(label) ;
+	}
+
+	HashType HashPart(const detail::Slice::PartProxy &part) {
+		HashType hash = 1;
+		for (const auto &label : part) {
+			HashType label_hash =
+			    (getLabelValue(label.first) * Goal::kMaxExpectedCopies +
+			     label.second) %
+			    kBasePart_;
+			hash = (hash * kBasePart_ + label_hash) % kModPart_;
+		}
+		return hash;
+	}
+
+	HashType HashSlice(HashType startingHash, Goal::Slice &slice) {
+		HashType hash = startingHash;
+		for (const auto &part : slice) {
+			HashType part_value = HashPart(part);
+			hash = (hash * kBaseSlice_ + part_value) % kModSlice_;
+		}
+		return hash;
+	}
+
+	template <std::size_t N>
+	bool checkMatchAndGetResult(Goal::Slice &target_slice,
+	                            Goal::Slice &source_slice,
+	                            std::array<int, N> &assignemt) {
+		HashType hash = HashSlice(kInitialStartingHash, target_slice);
+		hash = HashSlice(hash, source_slice);
+		if (cachedData_[hash].checkMatchAndGetResult<N>(target_slice, source_slice,
+		                                             assignemt)) {
+			return true;
+		}
+		return false;
+	}
+
+	template <std::size_t N>
+	void store(Goal::Slice &target_slice, Goal::Slice &source_slice,
+	           std::array<int, N> &assignemt) {
+		HashType hash = HashSlice(kInitialStartingHash, target_slice);
+		hash = HashSlice(hash, source_slice);
+		cachedData_[hash].store<N>(target_slice, source_slice, assignemt);
+	}
+
+protected:
+	const HashType kBasePart_ = 500009;
+	const HashType kModPart_ = 1000003;
+	const HashType kBaseSlice_ = 1500007;
+	const HashType kModSlice_ = 2000003;
+	const HashType kInitialStartingHash = 1;
+	EntriesContainer cachedData_;
```

Comment:
cachedData_ is currently a vector initialized here:
```cpp
LinearAssignmentCache() : cachedData_(kModSlice_) {}
```
kModSlice_ is 2000003, which is probably a lot of elements.

Please, consider to make it an unordered_map, like this:
```cpp
using EntriesContainer = std::unordered_map<HashType, OptimizerInputOutput>;
```
This way it will have only the needed elements and the lookup still will be O(1) in average.

Reply:
Applied.

At first I was a bit hesitant because I don't like too much unordered maps but it seems fair to trade the save of memory against some few very indexation extra operations (vs direct indexation).